README.md
Queue Server
============

Queue Server - by btnguyen2k.

Project home: [https://github.com/btnguyen2k/queue-server](https://github.com/btnguyen2k/queue-server).

## Overview ##

Queue service that unifies various queue backends into a single API set.

*Features:*

- Client-server model.
- API sets for Thrift, Thrift-over-HTTP and REST.
- Data format (for REST APIs): JSON
- Simple authorization scheme for API calls.

*Message format (JSON fields)*

- `queue_id`: (long) message's unique id in queue.
- `org_timestamp`: (long) UNIX timestamp (in milliseconds) when the message was first queued.
- `timestamp`: (long) UNIX timestamp (in milliseconds) when the message was queued.
- `num_requeues`: (int) number of times the messaged has been re-queued
- `content`: message's content in base64 encoding.


## Release-notes ##

Latest release: `0.3.1`.

See [RELEASE-NOTES.md](RELEASE-NOTES.md).


## APIs ##

_Each API call is authorized by an `authKey`._

* `initQueue(authkey, queueName)`: creates & initializes a new queue.
* `queue(authkey, queueName, message)`: put a message to a queue.
* `requeue(authkey, queueName, message)`: requeues a message (message's `num_requeues` will be increased).
* `requeueSilently(authkey, queueName, message)`: requeues a message "silently" (message's `num_requeues` will NOT be increased).
* `finish(authKey, queueName, message)`: called when finish processing the message to cleanup ephemeral storage.
* `take(authKey, queueName)`: takes a message from a queue.
* `queueExists(authKey, queueName)`: checks if a queue exists.
* `queueSize(authKey, queueName)`: gets number of items currently in a queue.
* `ephemeralSize(authKey, queueName)`: gets number of items currently in a queue's ephemeral storage.

### REST APIs ###

_All REST API calls must be made via POST method._

* `POST /initQueue`: 
  - Input: JSON `{"secret":"authorization key","queue_name":"name of the queue to be initialized"}`
  - Output: JSON `{"s":200,"r":true ,"m":"Successful: the queue has been created and initialized"}`
  - Output: JSON `{"s":200,"r":false,"m":"Successful: the queue has already existed"}`
  - Output: JSON `{"s":500,"r":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":403,"r":false,"m":"Error message: unauthorized request"}`
  - Output: JSON `{"s":400,"r":false,"m":"Error message: invalid parameters"}`

* `POST /queue`
  - Input: JSON `{"secret":"authorization key","queue_name":"name of the queue", "content":"message content in base64"}`
  - Output: JSON `{"s":200,"r":true ,"m":"Successful: the message has been queued"}`
  - Output: JSON `{"s":500,"r":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":404,"r":false,"m":"Error message: queue does not exist"}`
  - Output: JSON `{"s":403,"r":false,"m":"Error message: unauthorized request"}`
  - Output: JSON `{"s":400,"r":false,"m":"Error message: invalid parameters"}`
  
* `POST /requeue`
  - Input: JSON `{"secret":"authorization key","queue_name":"name of the queue","queue_id":(long),"org_timestamp":(long),"timestamp":(long),"num_requeues":(int),"content":"message content in base64"}`
  - Output: JSON `{"s":200,"r":true ,"m":"Successful: the message has been requeued"}`
  - Output: JSON `{"s":500,"r":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":404,"r":false,"m":"Error message: queue does not exist"}`
  - Output: JSON `{"s":403,"r":false,"m":"Error message: unauthorized request"}`
  - Output: JSON `{"s":400,"r":false,"m":"Error message: invalid parameters"}`

* `POST /requeueSilent`
  - Input: JSON `{"secret":"authorization key","queue_name":"name of the queue","queue_id":(long),"org_timestamp":(long),"timestamp":(long),"num_requeues":(int),"content":"message content in base64"}`
  - Output: JSON `{"s":200,"r":true ,"m":"Successful: the message has been requeued"}`
  - Output: JSON `{"s":500,"r":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":404,"r":false,"m":"Error message: queue does not exist"}`
  - Output: JSON `{"s":403,"r":false,"m":"Error message: unauthorized request"}`
  - Output: JSON `{"s":400,"r":false,"m":"Error message: invalid parameters"}`
  
* `POST /finish`
  - Input: JSON `{"secret":"authorization key","queue_name":"name of the queue","queue_id":(long),"org_timestamp":(long),"timestamp":(long),"num_requeues":(int),"content":"message content in base64"}`
  - Output: JSON `{"s":200,"r":true ,"m":"Successful"}`
  - Output: JSON `{"s":500,"r":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":404,"r":false,"m":"Error message: queue does not exist"}`
  - Output: JSON `{"s":403,"r":false,"m":"Error message: unauthorized request"}`
  - Output: JSON `{"s":400,"r":false,"m":"Error message: invalid parameters"}`

* `POST /take`
  - Input: JSON `{"secret":"authorization key","queue_name":"name of the queue"}`
  - Output: JSON `{"s":200,"r":true ,"m":"Successful","v":{"queue_id":(long),"org_timestamp":(long),"timestamp":(long),"num_requeues":(int),"content":"message's content in base64"}}`
  - Output: JSON `{"s":200,"r":false,"m":"Successful, but queue is empty"}`
  - Output: JSON `{"s":500,"r":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":404,"r":false,"m":"Error message: queue does not exist"}`
  - Output: JSON `{"s":403,"r":false,"m":"Error message: unauthorized request"}`
  - Output: JSON `{"s":400,"r":false,"m":"Error message: invalid parameters"}`

* `POST /queueExists`
  - Input: JSON `{"secret":"authorization key","queue_name":"name of the queue"}`
  - Output: JSON `{"s":200,"r":true ,"m":"true"}}`
  - Output: JSON `{"s":404,"r":false,"m":"false"}`
  - Output: JSON `{"s":500,"r":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":403,"r":false,"m":"Error message: unauthorized request"}`
  - Output: JSON `{"s":400,"r":false,"m":"Error message: invalid parameters"}`

* `POST /queueSize`
  - Input: JSON `{"secret":"authorization key","queue_name":"name of the queue"}`
  - Output: JSON `{"s":200,"r":true ,"v":(int)size}`
  - Output: JSON `{"s":200,"r":false,"v":-1,"m":"Operation is not supported"}`
  - Output: JSON `{"s":500,"r":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":404,"r":false,"m":"Error message: queue does not exist"}`
  - Output: JSON `{"s":403,"r":false,"m":"Error message: unauthorized request"}`
  - Output: JSON `{"s":400,"r":false,"m":"Error message: invalid parameters"}`

* `POST /ephemeralSize`
  - Input: JSON `{"secret":"authorization key","queue_name":"name of the queue"}`
  - Output: JSON `{"s":200,"r":true ,"v":(int)size}`
  - Output: JSON `{"s":200,"r":false,"v":-1,"m":"Operation is not supported"}`
  - Output: JSON `{"s":500,"r":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":404,"r":false,"m":"Error message: queue does not exist"}`
  - Output: JSON `{"s":403,"r":false,"m":"Error message: unauthorized request"}`
  - Output: JSON `{"s":400,"r":false,"m":"Error message: invalid parameters"}`


### Thrift APIs ###

See file [queueserver.thrift](thrift/queueserver.thrift).

### Thrift-over-HTTP APIs ###

Thrift APIs, but over HTTP(s)! URL: `http://server:host/thrift`

### Clients ###

- Java: [https://github.com/btnguyen2k/queue-jclient](https://github.com/btnguyen2k/queue-jclient).


## Installation ##

Note: Java 7+ is required!

### Install from binary ###

- Download binary package from [project release site](https://github.com/btnguyen2k/queue-server/releases).
- Unzip the binary package and copy it to your favourite location, e.g. `/usr/local/queue-server`.


### Install from source ###

- Download application's source, either cloning github project or download the source package from [project release site](https://github.com/btnguyen2k/queue-server/releases).
- Build [Play! Framework](https://www.playframework.com): `play dist`.
- The built binary package is available at `target/universal/queue-server-<version>.zip`. You may copy it to your favourite location, e.g. `/usr/local/queue-server`.


## Start/Stop Queue-Server ##

### Linux ###

Start server with default options:
> `/usr/local/queue-server/conf/server-production.sh start`

Start server with 1024M memory limit, REST & Thrift-over-HTTP APIs on port 18080, Thrift APIs on port 19090
> `/usr/local/queue-server/conf/server-production.sh start -m 1024 -p 18080 -t 19090`

Stop server:
> `/usr/local/queue-server/conf/server-production.sh stop`


## Configurations ##

### Start Script ###

Custom port numbers:
> ***-p port_number***: set port number for REST & Thrift-over-HTTP APIs. Example: *-m 18080*.

> ***-t port_number***: set port number for Thrift APIs. Example: *-t 19090*. Note: *-t 0* will disable Thrift APIs.

Memory limit:
> ***-m mem_in_mb***: set server's memory limit (unit: Megabytes). Example: *-m 1024*.

Custom configuration files:
> ***-c config_file***: set custom application configuration file, relative file is loaded under directory *${app.home}/conf*.
>
> Example *-c abc.conf*: use configuration file *${app.home}/conf/abc.conf*
>
> Example *-c /myapp/conf/abc.conf*: use configuration file */myapp/conf/abc.conf*

> ***-s config_file***: set custom spring configuration file, relative file is loaded under directory *${app.home}/conf*.
>
> Example *-c spring/beans.xml*: use configuration file *${app.home}/conf/spring/beans.xml*
>
> Example *-c /myapp/spring/beans.xml*: use configuration file */myapp/spring/beans.xml*

### SPring Configuration Files ###

Default file: `${app.home}/conf/spring/beans.xml`


## License ##

See [LICENSE.txt](LICENSE.txt) for details. Copyright (c) 2015-2016 btnguyen2k.

Third party libraries are distributed under their own license(s).


RELEASE-NOTES.md
Queue Server Release Notes
==========================

Queue Server - by btnguyen2k.

2016-02-25: v0.3.2
------------------

- NPE fix

2016-02-18: v0.3.1
------------------

- Support Redis backend (new class `RedisQueueApi`).
- Change `orphanMessageThresholdMs` to 60 seconds.
- Bugs fixed.


2015-06-22: v0.2.0
------------------

- Support PostgreSQL backend (new class `PgSQLQueueApi`).
- Start/Stop script updated: support custom application config & spring config files.
- Document review & fix.


2015-06-16: v0.1.0
------------------
First release:

- REST & Thrift APIs


app/globals/Registry.java
package globals;

import java.io.File;

import org.apache.commons.lang3.StringUtils;
import org.apache.thrift.server.TServer;
import org.springframework.beans.BeansException;
import org.springframework.context.ApplicationContext;
import org.springframework.context.support.AbstractApplicationContext;
import org.springframework.context.support.FileSystemXmlApplicationContext;

import play.Logger;
import play.Play;
import queue.QueueApi;

import com.github.ddth.tsc.DataPoint;
import com.github.ddth.tsc.ICounter;
import com.github.ddth.tsc.ICounterFactory;

public class Registry {

    public static void init() {
        initApplicationContext();

        localCounterFactory = getBean("TSC_LOCAL", ICounterFactory.class);
        globalCounterFactory = getBean("TSC_GLOBAL", ICounterFactory.class);
        queueApi = getBean(QueueApi.class);
    }

    public static void destroy() {
        stopThriftServer();

        destroyApplicationContext();
    }

    /*----------------------------------------------------------------------*/
    private static ICounterFactory localCounterFactory, globalCounterFactory;
    public final static String TSC_TOTAL = "BLOOMSERVER_TSC_TOTAL";
    public final static String TSC_200 = "BLOOMSERVER_TSC_200";
    public final static String TSC_400 = "BLOOMSERVER_TSC_400";
    public final static String TSC_403 = "BLOOMSERVER_TSC_403";
    public final static String TSC_404 = "BLOOMSERVER_TSC_404";
    public final static String TSC_500 = "BLOOMSERVER_TSC_500";

    public final static String COUNTER_TOTAL = "BLOOMSERVER_COUNTER_TOTAL";
    public final static String COUNTER_200 = "BLOOMSERVER_COUNTER_200";
    public final static String COUNTER_400 = "BLOOMSERVER_COUNTER_400";
    public final static String COUNTER_403 = "BLOOMSERVER_COUNTER_403";
    public final static String COUNTER_404 = "BLOOMSERVER_COUNTER_404";
    public final static String COUNTER_500 = "BLOOMSERVER_COUNTER_500";

    public final static String COUNTER_CONCURENCY = "BLOOMSERVER_CONCURENCY";

    public static void incConcurrency() {
        ICounter counter;

        counter = localCounterFactory != null ? localCounterFactory.getCounter(COUNTER_CONCURENCY)
                : null;
        if (counter != null) {
            counter.add(1000, 1);
        }

        counter = globalCounterFactory != null ? globalCounterFactory
                .getCounter(COUNTER_CONCURENCY) : null;
        if (counter != null) {
            counter.add(1000, 1);
        }
    }

    public static void decConcurrency() {
        ICounter counter;

        counter = localCounterFactory != null ? localCounterFactory.getCounter(COUNTER_CONCURENCY)
                : null;
        if (counter != null) {
            counter.add(1000, -1);
        }

        counter = globalCounterFactory != null ? globalCounterFactory
                .getCounter(COUNTER_CONCURENCY) : null;
        if (counter != null) {
            counter.add(1000, -1);
        }
    }

    public static long[] getConcurrency() {
        ICounter counter;
        DataPoint dp;
        long[] result = new long[2];

        counter = localCounterFactory != null ? localCounterFactory.getCounter(COUNTER_CONCURENCY)
                : null;
        dp = counter != null ? counter.get(1000) : null;
        result[0] = dp != null ? dp.value() : 0;

        counter = globalCounterFactory != null ? globalCounterFactory
                .getCounter(COUNTER_CONCURENCY) : null;
        dp = counter != null ? counter.get(1000) : null;
        result[1] = dp != null ? dp.value() : 0;

        return result;
    }

    private static void _updateTscCounters(final String name) {
        ICounter counterLocal = localCounterFactory != null ? localCounterFactory.getCounter(name)
                : null;
        if (counterLocal != null) {
            counterLocal.add(1);
        }

        ICounter counterGlobal = globalCounterFactory != null ? globalCounterFactory
                .getCounter(name) : null;
        if (counterGlobal != null) {
            counterGlobal.add(1);
        }
    }

    private static void _updateCounters(final String name) {
        ICounter counterLocal = localCounterFactory != null ? localCounterFactory.getCounter(name)
                : null;
        if (counterLocal != null) {
            counterLocal.add(0, 1);
        }

        ICounter counterGlobal = globalCounterFactory != null ? globalCounterFactory
                .getCounter(name) : null;
        if (counterGlobal != null) {
            counterGlobal.add(0, 1);
        }
    }

    public static void updateCounters(final int status) {
        _updateTscCounters(TSC_TOTAL);
        _updateCounters(COUNTER_TOTAL);
        switch (status) {
        case 200:
            _updateTscCounters(TSC_200);
            _updateCounters(COUNTER_200);
            break;
        case 400:
            _updateTscCounters(TSC_400);
            _updateCounters(COUNTER_400);
            break;
        case 403:
            _updateTscCounters(TSC_403);
            _updateCounters(COUNTER_403);
            break;
        case 404:
            _updateTscCounters(TSC_404);
            _updateCounters(COUNTER_404);
            break;
        case 500:
            _updateTscCounters(TSC_500);
            _updateCounters(COUNTER_500);
            break;
        }
    }

    public static ICounterFactory getLocalCounterFactory() {
        return localCounterFactory;
    }

    public static ICounterFactory getGlobalCounterFactory() {
        return globalCounterFactory;
    }

    /*----------------------------------------------------------------------*/
    private static QueueApi queueApi;

    public static QueueApi getQueueApi() {
        return queueApi;
    }

    /*----------------------------------------------------------------------*/
    private static TServer thriftServer = null;

    public static void startThriftServer(final TServer thriftServer) {
        Registry.thriftServer = thriftServer;
        Thread t = new Thread("Thrift Server") {
            public void run() {
                thriftServer.serve();
            }
        };
        t.start();
    }

    public static void stopThriftServer() {
        if (thriftServer != null) {
            try {
                thriftServer.stop();
            } catch (Exception e) {
                Logger.warn(e.getMessage(), e);
            } finally {
                thriftServer = null;
            }
        }
    }

    /*----------------------------------------------------------------------*/
    private static ApplicationContext applicationContext;

    public static <T> T getBean(Class<T> clazz) {
        try {
            return applicationContext.getBean(clazz);
        } catch (BeansException e) {
            return null;
        }
    }

    public static <T> T getBean(String name, Class<T> clazz) {
        try {
            return applicationContext.getBean(name, clazz);
        } catch (BeansException e) {
            return null;
        }
    }

    synchronized private static void initApplicationContext() {
        if (Registry.applicationContext == null) {
            String configFile = System.getProperty("spring.config.file", "conf/spring/beans.xml");
            if (!StringUtils.isEmpty(configFile)) {
                File springConfigFile = configFile.startsWith("/") ? new File(configFile)
                        : new File(Play.application().path(), configFile);
                AbstractApplicationContext applicationContext = new FileSystemXmlApplicationContext(
                        "file:" + springConfigFile.getAbsolutePath());
                applicationContext.start();
                Registry.applicationContext = applicationContext;
            }
        }
    }

    synchronized private static void destroyApplicationContext() {
        if (applicationContext != null) {
            try {
                ((AbstractApplicationContext) applicationContext).destroy();
            } catch (Exception e) {
                Logger.warn(e.getMessage(), e);
            } finally {
                applicationContext = null;
            }
        }
    }
}
